codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,319 @@
+            fireChildReadAndRegister(streamInfo, headers);
+            if (endOfStream) {
+                // TODO
+            }
+        }
+
+        @Override
+        public int onDataRead(ChannelHandlerContext ctx, int streamId, final ByteBuf data,
+                int padding, final boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            final StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            EventLoop childLoop = streamInfo.childChannel.eventLoop();
+            if (childLoop.inEventLoop()) {
+                childOnDataRead(streamInfo, data, endOfStream);
+            } else {
+                childLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        childOnDataRead(streamInfo, data, endOfStream);
+                    }
+                });
+            }
+            return data.readableBytes() + padding;
+        }
+
+        void childOnDataRead(StreamInfo streamInfo, ByteBuf data, boolean endOfStream) {
+            fireChildReadAndRegister(streamInfo, data);
+            if (endOfStream) {
+                // TODO
+            }
+        }
+    }
+
+    static class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    class Http2StreamChannel extends ChildChannel {
+        private final Http2Stream stream;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(Http2MultiplexCodec.this.ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        public void doClose() throws Exception {
+            EventLoop parentLoop = ctx.channel().eventLoop();
+            if (parentLoop.inEventLoop()) {
+                parentClose();
+            } else {
+                parentLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        parentClose();
+                    }
+                });
+            }
+            super.doClose();
+        }
+
+        public void parentClose() {
+            try {
+                if (stream.state() != Http2Stream.State.CLOSED) {
+                    http2Handler.resetStream(ctx, stream.id(), Http2Error.CANCEL.code(),
+                        ctx.newPromise());
+                }
+            } catch (Throwable t) {
+                parentExceptionCaught(t);
+            }
+        }
+
+        @Override
+        public void parentWrite(Object msg) throws Exception {
+            // TODO: fix endOfStream
```

Comment:
A user event is still an allocation, so there would be no advantage but with added complexity.  It would be much more Netty-ish (and understandable) to go with 1.